core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala

```diff
@@ -492,42 +492,59 @@ class GroupMetadataManager(brokerId: Int,
         val partitionData = new PartitionData(OffsetFetchResponse.INVALID_OFFSET,
           Optional.empty(), "", Errors.NONE)
-        topicPartition -> partitionData
+        topicIdPartition -> partitionData
       }.toMap
     } else {
       group.inLock {
         if (group.is(Dead)) {
-          topicPartitionsOpt.getOrElse(Seq.empty[TopicPartition]).map { topicPartition =>
+          topicIdPartitionsOpt.getOrElse(Seq.empty[TopicIdPartition]).map { topicIdPartition =>
             val partitionData = new PartitionData(OffsetFetchResponse.INVALID_OFFSET,
               Optional.empty(), "", Errors.NONE)
-            topicPartition -> partitionData
+            topicIdPartition -> partitionData
           }.toMap
         } else {
-          val topicPartitions = topicPartitionsOpt.getOrElse(group.allOffsets.keySet)
-
-          topicPartitions.map { topicPartition =>
-            if (requireStable && group.hasPendingOffsetCommitsForTopicPartition(topicPartition)) {
-              topicPartition -> new PartitionData(OffsetFetchResponse.INVALID_OFFSET,
+          def resolvePartitionData(topicIdPartition: TopicIdPartition): PartitionData = {
+            if (requireStable && group.hasPendingOffsetCommitsForTopicPartition(topicIdPartition)) {
+              new PartitionData(OffsetFetchResponse.INVALID_OFFSET,
                 Optional.empty(), "", Errors.UNSTABLE_OFFSET_COMMIT)
             } else {
-              val partitionData = group.offset(topicPartition) match {
+              group.offset(topicIdPartition) match {
                 case None =>
                   new PartitionData(OffsetFetchResponse.INVALID_OFFSET,
                     Optional.empty(), "", Errors.NONE)
                 case Some(offsetAndMetadata) =>
                   new PartitionData(offsetAndMetadata.offset,
                     offsetAndMetadata.leaderEpoch, offsetAndMetadata.metadata, Errors.NONE)
               }
-              topicPartition -> partitionData
             }
-          }.toMap
+          }
+
+          topicIdPartitionsOpt match {
+            case Some(topicIdPartitions) =>
+              topicIdPartitions.map { topicIdPartition =>
+                topicIdPartition -> resolvePartitionData(topicIdPartition)
+              }.toMap
+
+            case None =>
+              val topicIds = replicaManager.metadataCache.topicNamesToIds()
+              group.allOffsets.keySet.map { topicPartition =>
+                Option(topicIds.get(topicPartition.topic())) match {
+                  case Some(topicId) =>
+                    val topicIdPartition = new TopicIdPartition(topicId, topicPartition)
+                    topicIdPartition -> resolvePartitionData(topicIdPartition)
+                  case None =>
+                    val zeroIdPartition = new TopicIdPartition(Uuid.ZERO_UUID, topicPartition)
+                    zeroIdPartition -> OffsetFetchResponse.UNKNOWN_PARTITION
+                }
+              }.toMap
```

Comment:
I have been thinking about this part of the code and I don't think that we can treat it like this.

The first thing to note is that we hit this path when all partitions are requested. In this case, it feels weird to return a partition with a zero topic id and `OffsetFetchResponse.UNKNOWN_PARTITION`. I think that we should just ignore unknown partitions in this case. Note that partitions are deleted from the cache when the topic is deleted.

The second thing to note is that we can't ignore partitions with unknown topic ids all the time because the broker may be on an IBP which does not support topic ids. I think that we have to reason about it based on the version of the request.

The third thing, and this one is for the other patch, is that we should consider returning an UNSUPPORTED_VERSION if the last version of the API is used but the broker does not support topic ids, especially in the case where all offsets are requested.

All in all, I wonder if we should just return all offsets with zero topic ids here and handle all of that in KafkaApis. @Hangleton What do you think?

Reply:
It could if the IBP is kept on an old version.

Reply:
Hi David, thanks for taking the time to lay out these thoughts. I agree with you that using the zero id does not feel right. It can be considered as an abusive use of the `TopicIdPartition` as a reference to a resource in the cluster. In the worst case, an invalid `TopicIdPartition` could end up being used somewhere else which treats all `TopicIdPartition` as if they were valid/resolved. So, there is a clear code smell here.

In any case, the caller of the method `getOffsets` should know that some of the `TopicIdPartition` are invalid (that is, references the zero id). This defeats the purpose of returning a reference based on topic-id-partition. So, I agree with you that using topic-id-partition should not be exposed in the value returned by `getOffsets` even though there is an asymmetry in the method prototype as a result.

Reply:
Hi David and Justine, coming back to this, apologies for the delay. Are you comfortable of using `TopicIdPartition` as suggested above, or just keep `TopicPartition` and do the resolution to `TopicIdPartition` in `KafkaApis`? What do you think?

Reply:
I don't recall the details now. I have to get back to it...

Reply:
Sure that's fine! I will go ahead and start with the approach discussed last time.

Reply:
did we decide that topicIDPartition when the request is old would just have a 0 id? Could you give a brief outline for old vs new request versions and how they are handled (ie representation in memory when handling + what we return in the response for happy path and error cases)

Reply:
Sure, Justine. Will get back with a diagram.